File: java/src/gov/usgs/processingformats/TravelTimeRequest.java
package gov.usgs.processingformats;

import java.util.*;

import org.json.simple.JSONArray;
import org.json.simple.JSONObject;

import gov.usgs.processingformats.ProcessingInt;

/**
 * a conversion class used to create, parse, and validate travel time requests
 * 
 * @author U.S. Geological Survey &lt;jpatton at usgs.gov&gt;
 */
public class TravelTimeRequest implements ProcessingInt {

	/**
	 * JSON Keys
	 */
	public static final String TYPE_KEY = "Type";
	public static final String DISTANCE_KEY = "Distance";
	public static final String ELEVATION_KEY = "Elevation";
	public static final String LATITUDE_KEY = "Latitude";
	public static final String LONGITUDE_KEY = "Longitude";
	public static final String DATA_KEY = "Data";

	/**
	 * Required type of data as a String
	 */
	private String type;

	/**
	 * Required source receiver distance in seconds
	 */
	private Double distance;

	/**
	 * Required receiver elevation relative to the WGS84 datum in kilometers
	 */
	private Double elevation;

	/**
	 * Optional geographic receiver latitude in degrees
	 */
	private Double latitude;

	/**
	 * Optional geographic receiver longitude in degrees
	 */
	private Double longitude;

	/**
	 * Returned travel time data (empty for requests)
	 */
	private ArrayList<TravelTimeData> data;

	/**
	 * Returned travel time plot data (empty for requests)
	 */
	private ArrayList<TravelTimePlotData> plotData;

	/**
	 * The constructor for the TravelTimeData class. Initializes members to null
	 * values.
	 */
	public TravelTimeRequest() {

		reload(null, null, null, null, null, null, null);
	}

	/**
	 * Advanced constructor
	 * 
	 * The advanced constructor for the TravelTimeRequest class. Initializes
	 * members to provided values.
	 * 
	 * @param newType
	 *            - A String containing the request type, "Standard", "Plot", or
	 *            "PlotStatistics", defaults to standard
	 * @param newDistance
	 *            - A Double containing the source-receiver distance in degrees
	 * @param newElevation
	 *            - A Double containing the receiver elevation relative to the
	 *            WGS84 datum in kilometers
	 * @param newLatitude
	 *            - An optional Double containing the geographic receiver
	 *            latitude in degrees, null to omit
	 * @param newLongitude
	 *            - An optional Double containing the geographic receiver
	 *            longitude in degrees, null to omit
	 * @param newData
	 *            - A ArrayList&lt;TravelTimeData&gt; containing the returned
	 *            travel time data
	 * @param newPlotData
	 *            - A ArrayList&lt;TravelTimePlotData&gt;  containing the
	 *            returned travel time plot data
	 */
	public TravelTimeRequest(String newType, Double newDistance,
			Double newElevation, Double newLatitude, Double newLongitude,
			ArrayList<TravelTimeData> newData,
			ArrayList<TravelTimePlotData> newPlotData) {

		reload(newType, newDistance, newElevation, newLatitude, newLongitude,
				newData, newPlotData);
	}

	/**
	 * Constructs the class from a JSONObject, populating members
	 * 
	 * @param newJSONObject
	 *            - A JSONObject.
	 */
	public TravelTimeRequest(JSONObject newJSONObject) {

		// Required values
		// type
		if (newJSONObject.containsKey(TYPE_KEY)) {
			type = newJSONObject.get(TYPE_KEY).toString();
		} else {
			type = null;
		}

		// distance
		if (newJSONObject.containsKey(DISTANCE_KEY)) {
			distance = (double) newJSONObject.get(DISTANCE_KEY);
		} else {
			distance = null;
		}

		// elevation
		if (newJSONObject.containsKey(ELEVATION_KEY)) {
			elevation = (double) newJSONObject.get(ELEVATION_KEY);
		} else {
			elevation = null;
		}

		// latitude
		if (newJSONObject.containsKey(LATITUDE_KEY)) {
			latitude = (double) newJSONObject.get(LATITUDE_KEY);
		} else {
			latitude = null;
		}

		// longitude
		if (newJSONObject.containsKey(LONGITUDE_KEY)) {
			longitude = (double) newJSONObject.get(LONGITUDE_KEY);
		} else {
			longitude = null;
		}

		// Data
		if (newJSONObject.containsKey(DATA_KEY)) {

			data = new ArrayList<TravelTimeData>();
			plotData = new ArrayList<TravelTimePlotData>();

			// get the array
			JSONArray dataArray = (JSONArray) newJSONObject.get(DATA_KEY);

			if ((dataArray != null) && (!dataArray.isEmpty())) {

				// go through the whole array
				for (int i = 0; i < dataArray.size(); i++) {

					// get the object
					JSONObject dataObject = (JSONObject) dataArray.get(i);

					// check for type
					if (dataObject.containsKey(TYPE_KEY)) {

						// Route based on type
						String TypeString = (String) dataObject.get(TYPE_KEY);
						if (TypeString.equals("TTData")) {

							// add to vector
							data.add(new TravelTimeData(dataObject));
						} else if (TypeString.equals("TTPlotData")) {

							// add to vector
							plotData.add(new TravelTimePlotData(dataObject));
						}
					}
				}
			}
		} else {
			data = null;
			plotData = null;
		}
	}

	/**
	 * Constructs the class from a TravelTimeData object, populating members
	 * (copy constructor)
	 * 
	 * @param sourceObject
	 *            - A TravelTimeData object.
	 */
	public TravelTimeRequest(TravelTimeRequest sourceObject) {
		reload(sourceObject.type, sourceObject.distance, sourceObject.elevation,
				sourceObject.latitude, sourceObject.longitude,
				sourceObject.data, sourceObject.plotData);
	}

	/**
	 * Reload Function
	 * 
	 * The reload function for the TravelTimeData class. Initializes members to
	 * provided values.
	 * 
	 * @param newType
	 *            - A String containing the request type, "Standard", "Plot", or
	 *            "PlotStatistics", defaults to standard
	 * @param newDistance
	 *            - A Double containing the source-receiver distance in degrees
	 * @param newElevation
	 *            - A Double containing the receiver elevation relative to the
	 *            WGS84 datum in kilometers
	 * @param newLatitude
	 *            - An optional Double containing the geographic receiver
	 *            latitude in degrees, null to omit
	 * @param newLongitude
	 *            - An optional Double containing the geographic receiver
	 *            longitude in degrees, null to omit
	 * @param newData
	 *            - A ArrayList&lt;TravelTimeData&gt; containing the returned
	 *            travel time data
	 * @param newPlotData
	 *            - A ArrayList&lt;TravelTimePlotData&gt;  containing the
	 *            returned travel time plot data
	 */
	public void reload(String newType, Double newDistance, Double newElevation,
			Double newLatitude, Double newLongitude,
			ArrayList<TravelTimeData> newData,
			ArrayList<TravelTimePlotData> newPlotData) {

		if (newType == null) {
			type = "Standard";
		} else {
			type = newType;
		}
		distance = newDistance;
		elevation = newElevation;
		latitude = newLatitude;
		longitude = newLongitude;
		data = newData;
		plotData = newPlotData;
	}

	/**
	 * Converts the contents of the class to a json object
	 * 
	 * @return Returns a JSONObject containing the class contents
	 */
	@SuppressWarnings("unchecked")
	public JSONObject toJSON() {

		JSONObject newJSONObject = new JSONObject();

		String jsonType = getType();
		Double jsonDistance = getDistance();
		Double jsonElevation = getElevation();
		Double jsonLatitude = getLatitude();
		Double jsonLongitude = getLongitude();
		ArrayList<TravelTimeData> jsonData = getData();
		ArrayList<TravelTimePlotData> jsonPlotData = getPlotData();

		// required values
		// type
		if (jsonType != null) {
			newJSONObject.put(TYPE_KEY, jsonType);
		}

		// distance
		if (jsonDistance != null) {
			newJSONObject.put(DISTANCE_KEY, jsonDistance);
		}

		// elevation
		if (jsonElevation != null) {
			newJSONObject.put(ELEVATION_KEY, jsonElevation);
		}

		// optional values
		// latitude
		if (jsonLatitude != null) {
			newJSONObject.put(LATITUDE_KEY, jsonLatitude);
		}

		// longitude
		if (jsonLongitude != null) {
			newJSONObject.put(LONGITUDE_KEY, jsonLongitude);
		}

		// returned data
		JSONArray dataArray = new JSONArray();
		if ((jsonType != null) && (jsonType.equals("Standard"))) {

			// enumerate through the whole arraylist
			for (Iterator<TravelTimeData> dataIterator = jsonData
					.iterator(); dataIterator.hasNext();) {

				// convert pick to JSON object
				JSONObject dataObject = ((TravelTimeData) dataIterator.next())
						.toJSON();

				dataArray.add(dataObject);
			}
		} else if ((jsonType != null) && (jsonType.equals("Plot"))) {

			// enumerate through the whole arraylist
			for (Iterator<TravelTimePlotData> dataIterator = jsonPlotData
					.iterator(); dataIterator.hasNext();) {

				// convert pick to JSON object
				JSONObject dataObject = ((TravelTimePlotData) dataIterator
						.next()).toJSON();

				dataArray.add(dataObject);
			}

		} else if ((jsonType != null) && (jsonType.equals("PlotStatistics"))) {

			// enumerate through the whole arraylist
			for (Iterator<TravelTimePlotData> dataIterator = jsonPlotData
					.iterator(); dataIterator.hasNext();) {

				// convert pick to JSON object
				JSONObject dataObject = ((TravelTimePlotData) dataIterator
						.next()).toJSON();

				dataArray.add(dataObject);
			}
		}

		if (!dataArray.isEmpty()) {
			newJSONObject.put(DATA_KEY, dataArray);
		}

		return (newJSONObject);
	}

	/**
	 * Validates the class.
	 * 
	 * @return Returns true if successful
	 */
	public boolean isValid() {
		if (getErrors() == null) {
			return (true);
		} else if (getErrors().size() == 0) {
			return (true);
		} else {
			return (false);
		}
	}

	/**
	 * Gets any validation errors in the class.
	 * 
	 * @return Returns a List&lt;String&gt; of any errors found
	 */
	public ArrayList<String> getErrors() {
		ArrayList<String> errorList = new ArrayList<String>();

		String jsonType = getType();
		Double jsonDistance = getDistance();
		Double jsonElevation = getElevation();
		Double jsonLatitude = getLatitude();
		Double jsonLongitude = getLongitude();
		ArrayList<TravelTimeData> jsonData = getData();
		ArrayList<TravelTimePlotData> jsonPlotData = getPlotData();

		// required values
		// type
		if (jsonType == null) {
			// type not found
			errorList.add("No Type in TravelTimeRequest Class.");
		} else if (jsonType.isEmpty()) {
			// type empty
			errorList.add("Empty Type in TravelTimeRequest Class.");
		} else if (!(jsonType.equals("Standard") || jsonType.equals("Plot")
				|| jsonType.equals("PlotStatistics"))) {
			// wrong type
			errorList.add("Unsupported type in TravelTimeRequest Class.");
		}

		// distance
		if (jsonDistance == null) {
			// distance not found
			errorList.add("No Distance in TravelTimeRequest Class.");
		}

		// elevation
		if (jsonElevation == null) {
			// elevation not found
			errorList.add("No Elevation in TravelTimeRequest Class.");
		}

		// optional values
		// latitude
		if (jsonLatitude != null) {

			if ((jsonLatitude < -90) || (jsonLatitude > 90)) {
				// invalid latitude
				errorList.add(
						"Latitude in TravelTimeRequest not in the range of -90 to 90.");
			}
		}

		// longitude
		if (jsonLongitude != null) {

			if ((jsonLongitude < -180) || (jsonLongitude > 180)) {
				// invalid longitude
				errorList.add(
						"Longitude in TravelTimeRequest not in the range of -180 to 180.");
			}
		}

		// data
		// returned data
		if ((jsonType != null) && (jsonType.equals("Standard"))) {

			// data
			if (jsonData != null) {
				// enumerate through the whole arraylist
				for (Iterator<TravelTimeData> dataIterator = jsonData
						.iterator(); dataIterator.hasNext();) {

					// convert pick to JSON object
					TravelTimeData jsonTTData = ((TravelTimeData) dataIterator
							.next());

					if (!jsonTTData.isValid()) {
						errorList.add(
								"Invalid TravelTimeData in TravelTimeRequest Class of type Standard.");
						break;
					}
				}
				if ((jsonPlotData != null) && (!jsonPlotData.isEmpty())) {
					// wrong data
					errorList.add(
							"Erroneous TravelTimePlotData in TravelTimeRequest Class of type Standard.");
				}
			}

		} else if ((jsonType != null) && (jsonType.equals("Plot"))) {

			// plot data
			if (jsonPlotData != null) {
				// enumerate through the whole arraylist
				for (Iterator<TravelTimePlotData> dataIterator = jsonPlotData
						.iterator(); dataIterator.hasNext();) {

					// convert pick to JSON object
					TravelTimePlotData jsonTTPlotData = ((TravelTimePlotData) dataIterator
							.next());

					if (!jsonTTPlotData.isValid()) {
						errorList.add(
								"Invalid TravelTimePlotData in TravelTimeRequest Class of type Plot.");
						break;
					}
				}

				if ((jsonData != null) && (!jsonData.isEmpty())) {
					// wrong data
					errorList.add(
							"Erroneous TravelTimeData in TravelTimeRequest Class of type Plot.");
				}
			}
		} else if ((jsonType != null) && (jsonType.equals("PlotStatistics"))) {

			// enumerate through the whole arraylist
			for (Iterator<TravelTimePlotData> dataIterator = jsonPlotData
					.iterator(); dataIterator.hasNext();) {

				// convert pick to JSON object
				TravelTimePlotData jsonTTPlotData = ((TravelTimePlotData) dataIterator
						.next());

				if (!jsonTTPlotData.isValid()) {
					errorList.add(
							"Invalid TravelTimePlotData in TravelTimeRequest Class of type PlotStatistics.");
					break;
				}
			}

			if ((jsonData != null) && (!jsonData.isEmpty())) {
				// wrong data
				errorList.add(
						"Erroneous TravelTimeData in TravelTimeRequest Class of type PlotStatistics.");
			}
		}

		return (errorList);
	}

	/**
	 * @return the type
	 */
	public String getType() {
		return type;
	}

	/**
	 * @param type
	 *            the type to set
	 */
	public void setType(String type) {
		this.type = type;
	}

	/**
	 * @return the distance
	 */
	public Double getDistance() {
		return distance;
	}

	/**
	 * @param distance
	 *            the distance to set
	 */
	public void setDistance(Double distance) {
		this.distance = distance;
	}

	/**
	 * @return the elevation
	 */
	public Double getElevation() {
		return elevation;
	}

	/**
	 * @param elevation
	 *            the elevation to set
	 */
	public void setElevation(Double elevation) {
		this.elevation = elevation;
	}

	/**
	 * @return the latitude
	 */
	public Double getLatitude() {
		return latitude;
	}

	/**
	 * @param latitude
	 *            the latitude to set
	 */
	public void setLatitude(Double latitude) {
		this.latitude = latitude;
	}

	/**
	 * @return the longitude
	 */
	public Double getLongitude() {
		return longitude;
	}

	/**
	 * @param longitude
	 *            the longitude to set
	 */
	public void setLongitude(Double longitude) {
		this.longitude = longitude;
	}

	/**
	 * @return the data
	 */
	public ArrayList<TravelTimeData> getData() {
		return data;
	}

	/**
	 * @param data
	 *            the data to set
	 */
	public void setData(ArrayList<TravelTimeData> data) {
		this.data = data;
	}

	/**
	 * @return the plotData
	 */
	public ArrayList<TravelTimePlotData> getPlotData() {
		return plotData;
	}

	/**
	 * @param plotData
	 *            the plotData to set
	 */
	public void setPlotData(ArrayList<TravelTimePlotData> plotData) {
		this.plotData = plotData;
	}
}

File: cpp/src/TravelTimeRequest.cpp
#include <TravelTimeRequest.h>

#include <string>
#include <limits>
#include <exception>
#include <vector>

// JSON Keys
#define TYPE_KEY "Type"
#define DISTANCE_KEY "Distance"
#define ELEVATION_KEY "Elevation"
#define LATITUDE_KEY "Latitude"
#define LONGITUDE_KEY "Longitude"
#define DATA_KEY "Data"

namespace processingformats {

TravelTimeRequest::TravelTimeRequest() {
	type = "Standard";
	distance = std::numeric_limits<double>::quiet_NaN();
	elevation = std::numeric_limits<double>::quiet_NaN();
	latitude = std::numeric_limits<double>::quiet_NaN();
	longitude = std::numeric_limits<double>::quiet_NaN();
	data.clear();
	plotData.clear();
}

TravelTimeRequest::TravelTimeRequest(
		std::string newType, double newDistance, double newElevation,
		double newLatitude, double newLongitude,
		std::vector<TravelTimeData> newData,
		std::vector<TravelTimePlotData> newPlotData) {
	if (newType == "") {
		type = "Standard";
	} else {
		type = newType;
	}
	distance = newDistance;
	elevation = newElevation;
	latitude = newLatitude;
	longitude = newLongitude;
	data.clear();
	for (int i = 0; i < static_cast<int>(newData.size()); i++) {
		data.push_back(newData[i]);
	}
	plotData.clear();
	for (int i = 0; i < static_cast<int>(newPlotData.size()); i++) {
		plotData.push_back(newPlotData[i]);
	}
}

TravelTimeRequest::TravelTimeRequest(rapidjson::Value &json) {
	// required values
	// type
	if ((json.HasMember(TYPE_KEY) == true)
			&& (json[TYPE_KEY].IsString() == true)) {
		type = std::string(json[TYPE_KEY].GetString(),
							json[TYPE_KEY].GetStringLength());
	} else {
		type = "";
	}

	// distance
	if ((json.HasMember(DISTANCE_KEY) == true)
			&& (json[DISTANCE_KEY].IsNumber() == true)
			&& (json[DISTANCE_KEY].IsDouble() == true))
		distance = json[DISTANCE_KEY].GetDouble();
	else
		distance = std::numeric_limits<double>::quiet_NaN();

	// elevation
	if ((json.HasMember(ELEVATION_KEY) == true)
			&& (json[ELEVATION_KEY].IsNumber() == true)
			&& (json[ELEVATION_KEY].IsDouble() == true))
		elevation = json[ELEVATION_KEY].GetDouble();
	else
		elevation = std::numeric_limits<double>::quiet_NaN();

	// latitude
	if ((json.HasMember(LATITUDE_KEY) == true)
			&& (json[LATITUDE_KEY].IsNumber() == true)
			&& (json[LATITUDE_KEY].IsDouble() == true))
		latitude = json[LATITUDE_KEY].GetDouble();
	else
		latitude = std::numeric_limits<double>::quiet_NaN();

	// longitude
	if ((json.HasMember(LONGITUDE_KEY) == true)
			&& (json[LONGITUDE_KEY].IsNumber() == true)
			&& (json[LONGITUDE_KEY].IsDouble() == true))
		longitude = json[LONGITUDE_KEY].GetDouble();
	else
		longitude = std::numeric_limits<double>::quiet_NaN();

	// data
	data.clear();
	plotData.clear();
	if ((json.HasMember(DATA_KEY) == true)
			&& (json[DATA_KEY].IsArray() == true)) {
		rapidjson::Value dataarray;
		dataarray = json[DATA_KEY].GetArray();

		for (rapidjson::SizeType i = 0; i < dataarray.Size(); i++) {
			// parse
			rapidjson::Value & datavalue = dataarray[i];

			if ((datavalue.HasMember(TYPE_KEY) == true)
					&& (datavalue[TYPE_KEY].IsString() == true)) {
				std::string typeString = datavalue[TYPE_KEY].GetString();

				if (typeString == "TTData") {
					processingformats::TravelTimeData newData(datavalue);

					// add to vector
					data.push_back(newData);
				} else if (typeString == "TTPlotData") {
					processingformats::TravelTimePlotData newData(datavalue);

					// add to vector
					plotData.push_back(newData);
				}
			}
		}
	}
}

TravelTimeRequest::TravelTimeRequest(
		const TravelTimeRequest & newTravelTimeRequest) {
	type = newTravelTimeRequest.type;
	distance = newTravelTimeRequest.distance;
	elevation = newTravelTimeRequest.elevation;
	latitude = newTravelTimeRequest.latitude;
	longitude = newTravelTimeRequest.longitude;

	data.clear();
	for (int i = 0; i < static_cast<int>(newTravelTimeRequest.data.size());
			i++) {
		data.push_back(newTravelTimeRequest.data[i]);
	}

	plotData.clear();
	for (int i = 0; i < static_cast<int>(newTravelTimeRequest.plotData.size());
			i++) {
		plotData.push_back(newTravelTimeRequest.plotData[i]);
	}
}

TravelTimeRequest::~TravelTimeRequest() {
}

rapidjson::Value & TravelTimeRequest::toJSON(
		rapidjson::Value &json,
		rapidjson::MemoryPoolAllocator<rapidjson::CrtAllocator> &allocator) {
	json.SetObject();

	// required values
	// type
	rapidjson::Value typevalue;
	typevalue.SetString(rapidjson::StringRef(type.c_str()), allocator);
	json.AddMember(TYPE_KEY, typevalue, allocator);

	// distance
	if (std::isnan(distance) != true)
		json.AddMember(DISTANCE_KEY, distance, allocator);

	// elevation
	if (std::isnan(elevation) != true)
		json.AddMember(ELEVATION_KEY, elevation, allocator);

	// latitude
	if (std::isnan(latitude) != true)
		json.AddMember(LATITUDE_KEY, latitude, allocator);

	// longitude
	if (std::isnan(longitude) != true)
		json.AddMember(LONGITUDE_KEY, longitude, allocator);

	// data
	// build json array
	rapidjson::Value dataarray(rapidjson::kArrayType);
	if (type == "Standard") {
		if (data.size() > 0) {
			// build json array
			rapidjson::Value dataarray(rapidjson::kArrayType);

			for (int i = 0; i < static_cast<int>(data.size()); i++) {
				rapidjson::Value datavalue(rapidjson::kObjectType);
				data[i].toJSON(datavalue, allocator);
				dataarray.PushBack(datavalue, allocator);
			}
		}
	} else if (type == "Plot") {
		if (plotData.size() > 0) {
			// build json array
			rapidjson::Value dataarray(rapidjson::kArrayType);

			for (int i = 0; i < static_cast<int>(plotData.size()); i++) {
				rapidjson::Value datavalue(rapidjson::kObjectType);
				plotData[i].toJSON(datavalue, allocator);
				dataarray.PushBack(datavalue, allocator);
			}
		}
	} else if (type == "PlotStatistics") {
		if (plotData.size() > 0) {
			// build json array
			rapidjson::Value dataarray(rapidjson::kArrayType);

			for (int i = 0; i < static_cast<int>(plotData.size()); i++) {
				rapidjson::Value datavalue(rapidjson::kObjectType);
				plotData[i].toJSON(datavalue, allocator);
				dataarray.PushBack(datavalue, allocator);
			}
		}
	}

	if (dataarray.Size() > 0) {
		json.AddMember(DATA_KEY, dataarray, allocator);
	}

	return (json);
}

std::vector<std::string> TravelTimeRequest::getErrors() {
	std::vector<std::string> errorlist;

	// required data
	// type
	if (type == "") {
		// type empty
		errorlist.push_back("Empty Type in TravelTimeRequest Class.");
	} else if ((type != "Standard") || (type != "Plot")
			|| (type != "PlotStatistics")) {
		// wrong type
		errorlist.push_back("Unsupported type in TravelTimeRequest Class.");
	}

	// distance
	if (std::isnan(distance) == true) {
		// distance found
		errorlist.push_back("No Distance in TravelTimeRequest Class.");
	}

	// elevation
	if (std::isnan(elevation) == true) {
		// elevation not found
		errorlist.push_back("No Elevation in TravelTimeRequest Class.");
	}

	// optional data
	// latitude
	if (std::isnan(latitude) != true) {
		if ((latitude < -90) || (latitude > 90)) {
			errorlist.push_back("Invalid Latitude in TravelTimeRequest class.");
		}
	}

	// longitude
	if (std::isnan(longitude) != true) {
		if ((longitude < -180) || (longitude > 180)) {
			errorlist.push_back(
					"Invalid Longitude in TravelTimeRequest class.");
		}
	}

	if ((type != "") && (type == "Standard")) {
		if (data.size() > 0) {
			for (int i = 0; i < static_cast<int>(data.size()); i++) {
				if (data[i].isValid() == false) {
					errorlist.push_back(
							"Invalid TravelTimeData in TravelTimeRequest Class "
							"of type Standard.");
					break;
				}
			}
		}
	} else if ((type != "") && (type == "Plot")) {
		if (plotData.size() > 0) {
			for (int i = 0; i < static_cast<int>(plotData.size()); i++) {
				if (plotData[i].isValid() == false) {
					errorlist.push_back(
							"Invalid TravelTimeData in TravelTimeRequest Class "
							"of type Plot.");
					break;
				}
			}
		}
	} else if ((type != "") && (type == "PlotStatistics")) {
		if (plotData.size() > 0) {
			for (int i = 0; i < static_cast<int>(plotData.size()); i++) {
				if (plotData[i].isValid() == false) {
					errorlist.push_back(
							"Invalid TravelTimeData in TravelTimeRequest Class "
							"of type PlotStatistics.");
					break;
				}
			}
		}
	}

	return (errorlist);
}
}  // namespace processingformats
